### ISSUE TITLE
Minimal console app unhandled FileNotFoundException

### ISSUE THREAD

**Issue by mikehaas763** [dotnet/cli#291]
I installed the mac osx pkg installer from get.asp.net. I then created a very minimal hello world console app. I've ran `dotnet restore` which appears to have run successfully. Then when I run `dotnet run` I get this:

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

**Comment by mikehaas763** [comment]
project.json:
```json
{
    "version": "1.0.0-*",
    "compilationOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "System.Console": "4.0.0-beta-*",
        "System.Runtime": "4.0.21-beta-*"
    },

    "frameworks": {
        "dnxcore50": { }
    }
}
```

**Comment by adrian17** [comment]
Same issue on Debian Testing.

**Comment by blackdwarf** [comment]
The docs are now updated as well as the samples. Confirmed that fix works on Ubuntu. However, OS X still fails (El Capitan, 10.11.1) with the above issue. 

/cc @davidfowl @piotrpMSFT for visibility.

**Comment by piotrpMSFT** [comment]
@mikehaas, can you confirm @blackdwarf updates and let us know if this can be closed?

**Comment by mikehaas763** [comment]
@blackdwarf what are the doc updates you've made? Is it the change from manually creating a project.json file to just using `dotnet init` ?

**Comment by dlech** [comment]
Here are the changes: https://github.com/dotnet/core/commit/c61177243a0387832666e25d2a8d2bc64d43bedf

**Comment by mikehaas763** [comment]
@dlech thanks for pointing me to the change

@piotrpMSFT Yes I confirm that the docs were updated, but do you really want to have this bug closed before it's fixed when it's being referenced in the Beta-Jan milestone?

**Comment by jgowdy** [comment]
Still seems broken:

user@host:~/hello$ dotnet run

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

user@host:~/hello$ cat project.json
{
    "version": "1.0.0-*",
    "compilationOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "Microsoft.NETCore.Runtime": "1.0.1-beta-*",
        "System.IO": "4.0.11-beta-23516",
        "System.Console": "4.0.0-beta-*",
        "System.Runtime": "4.0.21-beta-*"
    },

    "frameworks": {
        "dnxcore50": { }
    }
}

**Comment by blackdwarf** [comment]
Is this on os x @jgowdy?

**Comment by jgowdy** [comment]
Ubuntu 15.04

**Comment by brthor** [comment]
@jgowdy Can you give this a try before `dotnet run` ?

`dotnet restore -s https://myget.org/f/dotnet-core`

**Comment by jgowdy** [comment]
I just reinstalled with Ubuntu 14.04 LTS and can no longer reproduce the problem after changing System.IO to "4.0.11-beta-*"  Specifically targeting 23516 wasn't needed.

It's possible the cruft from one of my previous dnvm / dnx setups was involved.  I'm going to reinstall 15.04 to see if I can reproduce it version specific.

**Comment by jgowdy** [comment]
OSX 10.11.1 El Capitan:

Jeremiahs-MacBook-Pro:hello jgowdy$ dotnet run

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

**Comment by piotrpMSFT** [comment]
Ok, yea, we will take a look.

**Comment by anurse** [comment]
This is all due to our extremely minimal support of RIDs (see https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.ProjectModel/RuntimeIdentifier.cs and https://github.com/dotnet/cli/blob/master/scripts/_common.sh#L68)

We need to port the runtime identification code DNX has over, but that was heavily P/Invoke dependent. We are hoping to get an API in CoreFX we can use (https://github.com/dotnet/cli/issues/185#issuecomment-156205814)

/cc @davidfowl

**Comment by adrian17** [comment]
For impatient Ubuntu non-LTS1 users, there is an ugly unsafe temporary workaround: edit `/etc/os-release` and change the following values:

    ID=ubuntu
    VERSION_ID="14.04"

Then rerun `dotnet restore` (remember to change `os-release` back to its original version immediately afterwards!). After this, `dotnet run` properly runs a minimal console application. I tested it on Ubuntu 15.10, 15.04 and Debian.

**Comment by TomPoulton** [comment]
I'm seeing this on Mac OSX as well (El Capitan 10.11.1)

```
$ dotnet init
Created new project in /Users/... etc ...
$ ls
Program.cs	project.json
$ dotnet restore
Microsoft .NET Development Utility CoreClr-x64-1.0.0-rc1-16048

  CACHE https://www.myget.org/F/aspnetvnext/api/v2/
  CACHE https://www.nuget.org/api/v2/
Restoring packages for /Users/tom/... etc ...
... etc ...
$ dotnet run

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
```

**Comment by davidfowl** [comment]
Yes El Capitan restore is broken. To work around it you have to do `dotnet restore osx.10.10-x64`

**Comment by davidfowl** [comment]
@anurse We need to fix this.

**Comment by TomPoulton** [comment]
I think you meant `dotnet restore --runtime osx.10.10-x64` ;) but yes that worked!

**Comment by rogah** [comment]
El Capitan (10.11.1) restore is broken. The work around worked for now but it's frustrating though.

**Comment by anurse** [comment]
I'm actively working on improving the OS detection by porting some DNX code. Probably not going to be fully integrated until early next week, but it's on the way :).

**Comment by jgowdy** [comment]
@anurse Excellent, thank you.

**Comment by brthor** [comment]
@anurse Nice! :smile:

**Comment by TomPoulton** [comment]
This is such a surreal experience, conversing with Microsoft on GitHub about a .NET bug on Mac OSX! Love it :)
Thanks for the quick turnaround @anurse

**Comment by blackdwarf** [comment]
@anurse is this the same as #498? Should we close this one?

**Comment by anurse** [comment]
Kinda. They both refer to the same problem: We need accurate RID detection. Neither really captures that fully. Maybe create a new bug with that and dupe these to that bug.

**Comment by ochowie** [comment]
@anurse Will this fix the issue on Ubuntu 15.04. I can confirm that the temporary fix @adrian17 proposed works but it's really not a sustainable solution.

**Comment by anurse** [comment]
It should

**Comment by blackdwarf** [comment]
@anurse is the change you mention above done?

**Comment by VivekRagunathan** [comment]
Same error here:

```
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
```

I tried setting this: `DOTNET_HOME=/usr/local/share/dotnet/` too as suggested in one of the forums.

No solution yet? Any workarounds?

**Comment by dlech** [comment]
Please read/search the previous comments. Someone is working on it, but it is not fixed yet. There are several workarounds that have been mentioned, including one for El Capitan.

**Comment by VivekRagunathan** [comment]
Yes I already did. I am running on OSX 10.11.2, and the `dotnet restore --runtime osx.10.10-x64` does not work its magic. I guess I will have to wait for the real fix.

**Comment by dlech** [comment]
FWIW, I just updated to OS X 10.11.2 and `dotnet restore --runtime osx.10.10-x64` still works for me.

**Comment by anurse** [comment]
The pre-requisite change in the PlatformAbstractions library is done, but I wasn't able to get the necessary changes made to dotnet-cli before leaving for the holidays.

**Comment by blackdwarf** [comment]
@anurse thanks for the update! :)

**Comment by simranjitr** [comment]
I did the following (on OSX 10.11.2) to get the basic 'dotnet new' generated program up and running:

##########
export DOTNET_HOME=/usr/local/share/dotnet/
##########

###project.json####
{
    "version": "1.0.0-*",
    "compilationOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "System.Console": "4.0.0-beta-23409", 
        "System.IO": "4.0.11-beta-*",
        "System.Runtime": "4.0.21-beta-*" 
    },

    "frameworks": {
        "dnxcore50": { }
    }
}


$dotnet restore --runtime osx.10.10-x64
$dotnet run


which worked.

**Comment by anurse** [comment]
Yeah, that's because `dotnet-restore` knows about `osx.10.11` and so it puts that in the lock file, but `dotnet-run` does **not**, so it's looking for `osx.10.10` :). It's exactly the issue I'm working on resolving in #498

**Comment by anurse** [comment]
Should be fixed when #498 is closed. If not, please feel free to reopen!

**Comment by loic-sharma** [comment]
Hey, seems like this bug is back:

```
$ dotnet new
Created new project in  /Users/sharmaloic/Code/REPL
$ dotnet restore
Microsoft .NET Development Utility CoreClr-x64-1.0.0-rc1-16231

  CACHE https://api.nuget.org/v3/index.json
Restoring packages for /Users/sharmaloic/Code/REPL/project.json
  GET https://api.nuget.org/v3-flatcontainer/microsoft.netcore.runtime/index.json
  GET https://api.nuget.org/v3-flatcontainer/system.io/index.json
  GET https://api.nuget.org/v3-flatcontainer/system.console/index.json
  GET https://api.nuget.org/v3-flatcontainer/system.runtime/index.json
  OK https://api.nuget.org/v3-flatcontainer/system.io/index.json 300ms
  OK https://api.nuget.org/v3-flatcontainer/microsoft.netcore.runtime/index.json 525ms
  OK https://api.nuget.org/v3-flatcontainer/system.console/index.json 387ms
  OK https://api.nuget.org/v3-flatcontainer/system.runtime/index.json 433ms
Writing lock file /Users/sharmaloic/Code/REPL/project.lock.json
Restore complete, 7195ms elapsed

NuGet Config files used:
    /Users/sharmaloic/.config/NuGet/nuget.config

Feeds used:
    https://api.nuget.org/v3-flatcontainer/
$ dotnet run

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
$ dotnet build

Unhandled Exception: System.ComponentModel.Win32Exception: No such file or directory
   at System.Diagnostics.Process.ResolvePath(String filename)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at Microsoft.DotNet.Cli.Utils.Command.Execute()
   at Microsoft.DotNet.Cli.Program.Main(String[] args)
Abort trap: 6
$ dotnet restore --runtime osx.10.10-x64
   ....
$ dotnet run
Hello World!
```

I'm on El Capitan 10.11.

**Comment by salvarubiomartinez** [comment]
Same thing happen to me

**Comment by BrendanBuono** [comment]
Seeing the same with on .NET Core 1.0 Mac OS X

**Comment by VivekRagunathan** [comment]
Check all dependencies - openssl etc. Also check the restore command for osx . Alternatively you can try the same with vscode to make sure all deps are ok.

**Comment by dlech** [comment]
This is still an issue for me on macOS Sierra, following the hello world example from <https://www.microsoft.com/net/core#macos>.

Following the steps provided in the link, (`dotnet new; dotnet restore; dotnet run`) I get the exception as described in the first comment of this issue when trying to run with `dotnet run`. The `dotnet restore --runtime osx.10.10-x64` workaround is still required in order to run without throwing an exception.

I assume that I am using the `dotnet` command from `dotnet-dev-osx-x64.1.0.0-preview2-1-003177.pkg` that I just installed, but there is no `dotnet --version` command, so I don't know if I am actually using that version or an old version of `dotnet` that I installed a year ago.

**Comment by blackdwarf** [comment]
@dletch we do have the `--version` option in the CLI. When you say you don't uave it what do you mean by that?

**Comment by dlech** [comment]
I'm guessing that I have two versions of `dotnet` installed, one that I installed a year ago and one that I installed today. I haven't been using it in between. The older version must be in `$PATH` before the newer version. If I run `dotnet --version`, I get `Unknown option: --version`.

So, I imagine this bug is fixed and I am just seeing it still because I am running a very old `dotnet` that I need to figure out how to uninstall so that it doesn't get in the way of the current version.
